{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OpenSNP data formatting and analysis\n",
    "#### Adam Klie<br>05/20/2020<br>CSE284 Project\n",
    "Notebook to look at genotypes and phenotypes in OpenSNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "from tqdm._tqdm_notebook import tqdm_notebook\n",
    "tqdm_notebook.pandas()\n",
    "from scipy.stats import zscore"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Retrieve user IDs with eye color phenotype data from OpenSNP\n",
    " - Input: \n",
    "     - `phenotypes_202004220659.csv` -- phenotype data for downloaded openSNP genotype files\n",
    " - Output: \n",
    "     - `openSNP_initial_phenotypes.tab` -- a tabular version of initial set of users with phenotype data (638)\n",
    "     - `openSNP_initial_userid.list` -- list of user IDs (one per line) for those users with eye color phenotype data\n",
    " - Notes:\n",
    "     - Only looking at individuals with the following eye colors: 'Brown', 'brown', 'Blue', 'blue', 'Green', 'green'\n",
    "     - In the following columns: \"Eye color\", \"Eye Color\", \"Eye pigmentation \"\n",
    "     - 638 such individuals come out of this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in OpenSNP phenotypes\n",
    "phenotypes = pd.read_csv('phenotypes_202004220659.csv', sep=';')[['user_id', 'genotype_filename', 'date_of_birth', \n",
    "                                                                  'chrom_sex', 'Eye color', 'Eye Color', \n",
    "                                                                  'Eye pigmentation ']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove non-reporters\n",
    "phenotypes = phenotypes[(phenotypes[\"Eye color\"] != '-') | (phenotypes[\"Eye Color\"] != '-') | (phenotypes[\"Eye pigmentation \"] != '-')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Uncomment to look at useful phenotype values\n",
    "#phenotypes[\"Eye color\"].value_counts()\n",
    "#phenotypes[\"Eye Color\"].value_counts()\n",
    "#phenotypes[\"Eye pigmentation \"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only using these phenotype labels\n",
    "valid_phenotypes = ['Brown', 'brown', 'Blue', 'blue', 'Green', 'green']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only take those samples that have valid phenotypes in one of these three columns\n",
    "phenotypes = phenotypes[(phenotypes[\"Eye color\"].isin(valid_phenotypes)) | \n",
    "                        (phenotypes[\"Eye Color\"].isin(valid_phenotypes)) | \n",
    "                        (phenotypes[\"Eye pigmentation \"].isin(valid_phenotypes))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collapes to a single column\n",
    "def collapseEyeColors(x, values):\n",
    "    if x[\"Eye color\"] not in values:\n",
    "        if x[\"Eye Color\"] in values:\n",
    "            return x[\"Eye Color\"].lower()\n",
    "        elif x[\"Eye pigmentation \"] in values:\n",
    "            return x[\"Eye pigmentation \"].lower()\n",
    "    else:\n",
    "        return x[\"Eye color\"].lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use function to collapse eye color metadata to single columns\n",
    "phenotypes[\"eye_color\"] = phenotypes.apply(lambda x: collapseEyeColors(x, valid_phenotypes), 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove duplicate individuals\n",
    "phenotypes = phenotypes.drop_duplicates(subset=\"user_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only take Ancestry and 23andMe data to simplify\n",
    "phenotypes = phenotypes[phenotypes[\"genotype_filename\"].str.contains('ancestry') | phenotypes[\"genotype_filename\"].str.contains('23andme')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "brown    333\n",
      "blue     208\n",
      "green     97\n",
      "Name: eye_color, dtype: int64\n",
      "638\n"
     ]
    }
   ],
   "source": [
    "print(phenotypes[\"eye_color\"].value_counts())\n",
    "print(phenotypes[\"eye_color\"].value_counts().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotypes.to_csv('openSNP_initial_phenotypes.tab', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save this id set as a list\n",
    "ids = set(phenotypes[\"user_id\"].values)\n",
    "with open(\"openSNP_initial_userid.list\", 'w') as filehandle:\n",
    "    filehandle.writelines(\"%s\\n\" % id for id in ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Extracting initial genotypes using user_ids with eye_color phenotypes and filtering for high coverage SNPS\n",
    "**Run `extractGenotypes.py` to generate a tsv of rsids vs genotypes (`openSNP_initial_genotypes.tab`)**<br>\n",
    " - Input: \n",
    "     - `openSNP_initial_genotypes.tab` -- genotype data for downloaded openSNP genotype files\n",
    " - Output: \n",
    "     - `openSNP_filtered_genotypes.tab` -- a tabular version of genotypes filtered\n",
    "     - `openSNP_filtered_rsids.list` -- list of rsids to pass to 1000Genomes to find intersection with\n",
    " - Notes:\n",
    "     - Filter this dataframe to bring the number of SNPs down (currently at least 80% of samples have rsid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# AncestryDNA example data\n",
    "#!head -20 ~/cse284/datasets/openSNP/genotypes/user1001_file496_yearofbirth_unknown_sex_unknown.ancestry.txt\n",
    "#!wc -l ~/cse284/datasets/openSNP/genotypes/user1001_file496_yearofbirth_unknown_sex_unknown.ancestry.txt\n",
    "\n",
    "# 23andMe example data\n",
    "#!head -20 ~/cse284/datasets/openSNP/genotypes/user1002_file497_yearofbirth_unknown_sex_unknown.23andme.txt\n",
    "#!wc -l ~/cse284/datasets/openSNP/genotypes/user1002_file497_yearofbirth_unknown_sex_unknown.23andme.txt\n",
    "\n",
    "# FTDNA example data\n",
    "#!head -20 ~/cse284/datasets/openSNP/genotypes/user1018_file3018_yearofbirth_unknown_sex_XY.ftdna-illumina.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (161,183,242,376) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "# Read in genotypes\n",
    "initial_genotypes = pd.read_csv('openSNP_initial_genotypes.tab', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "num_SNPs = len(initial_genotypes)\n",
    "num_samples = len(initial_genotypes.loc[:, \"6\":\"6131\"].columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After extracting genotypes we have 1269707 SNP x 593 sample matrix\n"
     ]
    }
   ],
   "source": [
    "print(\"After extracting genotypes we have {} SNP x {} sample matrix\".format(num_SNPs, num_samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make sure we have iris SNPs captured\n",
    "IRISPLEX = os.path.join(os.environ[\"HOME\"], \"project/datasets/\", \"irisplex.bed\")\n",
    "iris_header = [\"chr\", \"pos1\", \"pos2\", \"id\", \"minor_allele\", \"b1\", \"b2\"]\n",
    "iris = pd.read_csv(IRISPLEX, sep='\\t', header=None, names=iris_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iris_snps = iris[\"id\"].values\n",
    "iris_gts = initial_genotypes[initial_genotypes[\"rsid\"].isin(iris_snps)]\n",
    "min(iris_gts.notna().apply(np.mean, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for SNPs that appear in at least 80% of samples (> than min iris SNP fraction)\n",
    "common_snp_genotypes = genotypes[genotypes.loc[:,'6':].notna().apply(np.mean, 1) > 0.80]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write this filtered set to a smaller dataframe\n",
    "common_snp_genotypes.to_csv('openSNP_filtered_genotypes.tab', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save this list of SNP rsids for oneK_genotpyes.sh\n",
    "final_rsids = common_snp_genotypes[\"rsid\"].values.tolist()\n",
    "with open(\"openSNP_filtered_rsids.list\", 'w') as filehandle:\n",
    "    filehandle.writelines(\"%s\\n\" % id for id in final_rsids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"After filtering, %d SNPs and %d samples left\" % (len(final_rsids), len(user_id_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Get final genotypes from filtered file and output of oneK genomes\n",
    " - Input: \n",
    "     - `openSNP_filtered_genotypes.tab` -- genotype data for downloaded openSNP genotype files\n",
    "     - `../oneKGenomes/oneK_rsids.tab` -- see how to build this in ../oneKGenomes\n",
    " - Output: \n",
    "     - `openSNP_filtered_genotypes.tab` -- a tabular version of genotypes filtered\n",
    "     - `openSNP_filtered_rsids.list` -- list of rsids to pass to 1000Genomes to find intersection with\n",
    " - Notes:\n",
    "     - Comes from final SNP list from 1K genomes (run oneK_genotypes.sh to get oneK_rsids.tab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (376) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "# Read in filtered genotypes\n",
    "filtered_genotypes = pd.read_csv('openSNP_filtered_genotypes.tab', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the final SNP the final SNP set\n",
    "oneK_ref = pd.read_csv('../oneKGenomes/oneK_rsids.tab', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace -- with NaN\n",
    "filtered_genotypes = filtered_genotypes.replace('--', np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only SNPs found in 1000Genomes\n",
    "gt_char = filtered_genotypes.merge(oneK_ref, left_on=\"rsid\", right_on=\"ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>6</th>\n",
       "      <th>8</th>\n",
       "      <th>2056</th>\n",
       "      <th>4106</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>13</th>\n",
       "      <th>...</th>\n",
       "      <th>6117</th>\n",
       "      <th>2023</th>\n",
       "      <th>2024</th>\n",
       "      <th>6120</th>\n",
       "      <th>6124</th>\n",
       "      <th>4080</th>\n",
       "      <th>8178</th>\n",
       "      <th>6131</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs3131972</td>\n",
       "      <td>1</td>\n",
       "      <td>742584</td>\n",
       "      <td>AG</td>\n",
       "      <td>AG</td>\n",
       "      <td>AG</td>\n",
       "      <td>GG</td>\n",
       "      <td>AG</td>\n",
       "      <td>GG</td>\n",
       "      <td>AG</td>\n",
       "      <td>...</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>AG</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>rs3131972</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs4970383</td>\n",
       "      <td>1</td>\n",
       "      <td>828418</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AC</td>\n",
       "      <td>AC</td>\n",
       "      <td>AC</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>AC</td>\n",
       "      <td>CC</td>\n",
       "      <td>AC</td>\n",
       "      <td>CC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC</td>\n",
       "      <td>rs4970383</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs4475691</td>\n",
       "      <td>1</td>\n",
       "      <td>836671</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>CT</td>\n",
       "      <td>CT</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>rs4475691</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs7537756</td>\n",
       "      <td>1</td>\n",
       "      <td>844113</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AG</td>\n",
       "      <td>AG</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>rs7537756</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs13302982</td>\n",
       "      <td>1</td>\n",
       "      <td>851671</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>...</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>rs13302982</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 598 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         rsid  chromosome  position   6   8 2056 4106  10  11  13 ... 6117  \\\n",
       "0   rs3131972           1    742584  AG  AG   AG   GG  AG  GG  AG ...   GG   \n",
       "1   rs4970383           1    828418  CC  CC   AA   AC  AC  AC  CC ...   CC   \n",
       "2   rs4475691           1    836671  CC  CC   CC   CT  CT  CT  CC ...   CC   \n",
       "3   rs7537756           1    844113  AA  AA   AA   AG  AG  AG  AA ...   AA   \n",
       "4  rs13302982           1    851671  GG  GG   GG   GG  GG  GG  GG ...   GG   \n",
       "\n",
       "  2023 2024 6120 6124 4080 8178 6131          ID REF  \n",
       "0   GG   GG   GG   AG   AG   AA   GG   rs3131972   A  \n",
       "1   CC   AC   CC   AC   CC  NaN   CC   rs4970383   C  \n",
       "2   CC   CT   CC   CT   CC   CC   CC   rs4475691   C  \n",
       "3   AA   AG   AA   AG   AA   AA   AA   rs7537756   A  \n",
       "4   GG   GG   GG   GG   GG   AA   GG  rs13302982   A  \n",
       "\n",
       "[5 rows x 598 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gt_char.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to convert genotype from nucleotides to numbers 0 (homozygous reference), 1(heteroozygous), 2 (homozygous minor allele)\n",
    "# Alos fixes any SNPs with -- to NaN\n",
    "def get_gt(x):\n",
    "    for col in x.index[3:-2]:       \n",
    "        if isinstance(x[col], str):\n",
    "            ref_allele = x[\"REF\"]\n",
    "            gt = 2 - x[col].upper().count(ref_allele)\n",
    "            x[col] = gt\n",
    "    return x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1bfeb9e241204256a02ec48ff5a9bcba",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=276549.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# WARNING: Getting the numbers takes about an hour and half\n",
    "gt_nums = gt_char.progress_apply(get_gt, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save a final SNP vs indiviudal table with other SNP info in it\n",
    "gt_nums = gt_nums.drop(\"ID\", axis=1)\n",
    "gt_nums.to_csv(\"openSNP_final_genotypes.tab\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to z-score SNPs\n",
    "def z_score(df): \n",
    "    return (df-df.mean())/df.std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "317cbb4f3bc84d74b26f4261f30b299e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=276549.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Z-score SNPs\n",
    "z_scored_snps = gt_nums.set_index(\"rsid\").loc[:, \"6\":\"6131\"].progress_apply(z_score, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = z_scored_snps.T.loc[phenotypes[phenotypes[\"user_id\"].isin(test.index)][\"user_id\"].astype(str).tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = test.replace(np.nan, 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "test.to_csv(\"test_set.csv\", index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_ids = test.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Final user id list\n",
    "with open(\"openSNP_final_userid.list\", 'w') as filehandle:\n",
    "    filehandle.writelines(\"%s\\n\" % id for id in test_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Get final phenotypes from..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"openSNP_final_userid.list\", 'r') as filename:\n",
    "    final_user_ids = [line.rstrip() for line in filename.readlines()]\n",
    "phenos = pd.read_csv('openSNP_initial_phenotypes.tab', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the phenotypes\n",
    "final_phenotypes = phenos[phenos[\"user_id\"].isin(final_user_ids)]\n",
    "final_phenotypes[[\"user_id\", \"eye_color\"]].to_csv('openSNP_final_phenotypes.tab', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "brown    312\n",
       "blue     194\n",
       "green     84\n",
       "Name: eye_color, dtype: int64"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_phenotypes[\"eye_color\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_phenotypes = final_phenotypes.set_index(\"user_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genotype_filename</th>\n",
       "      <th>date_of_birth</th>\n",
       "      <th>chrom_sex</th>\n",
       "      <th>Eye color</th>\n",
       "      <th>Eye Color</th>\n",
       "      <th>Eye pigmentation</th>\n",
       "      <th>eye_color</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>user_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4440</th>\n",
       "      <td>4440.ancestry.3043</td>\n",
       "      <td>1954</td>\n",
       "      <td>XX</td>\n",
       "      <td>Brown</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>brown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4441</th>\n",
       "      <td>4441.23andme.3044</td>\n",
       "      <td>rather not say</td>\n",
       "      <td>rather not say</td>\n",
       "      <td>Blue</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>blue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4106</th>\n",
       "      <td>4106.23andme.2779</td>\n",
       "      <td>rather not say</td>\n",
       "      <td>rather not say</td>\n",
       "      <td>Blue</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>blue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4171</th>\n",
       "      <td>4171.23andme.2830</td>\n",
       "      <td>rather not say</td>\n",
       "      <td>rather not say</td>\n",
       "      <td>Brown</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>brown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3973</th>\n",
       "      <td>3973.23andme.2669</td>\n",
       "      <td>1992</td>\n",
       "      <td>XY</td>\n",
       "      <td>Brown</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>brown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          genotype_filename   date_of_birth       chrom_sex Eye color  \\\n",
       "user_id                                                                 \n",
       "4440     4440.ancestry.3043            1954              XX     Brown   \n",
       "4441      4441.23andme.3044  rather not say  rather not say      Blue   \n",
       "4106      4106.23andme.2779  rather not say  rather not say      Blue   \n",
       "4171      4171.23andme.2830  rather not say  rather not say     Brown   \n",
       "3973      3973.23andme.2669            1992              XY     Brown   \n",
       "\n",
       "        Eye Color Eye pigmentation  eye_color  \n",
       "user_id                                        \n",
       "4440            -                 -     brown  \n",
       "4441            -                 -      blue  \n",
       "4106            -                 -      blue  \n",
       "4171            -                 -     brown  \n",
       "3973            -                 -     brown  "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_phenotypes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use numeric labels for phenotypes\n",
    "label_mapping = {\"brown\":0, \"blue\":1, \"green\":2}\n",
    "def get_label(x):\n",
    "    return label_mapping[x[\"eye_color\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_phenotypes[\"label\"] = final_phenotypes.apply(get_label, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_phenotypes[\"label\"].to_csv('test_labels.csv', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test with iris"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos1</th>\n",
       "      <th>pos2</th>\n",
       "      <th>id</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>b1</th>\n",
       "      <th>b2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15</td>\n",
       "      <td>28365618</td>\n",
       "      <td>28365619</td>\n",
       "      <td>rs12913832</td>\n",
       "      <td>A</td>\n",
       "      <td>-4.81</td>\n",
       "      <td>-1.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15</td>\n",
       "      <td>28230318</td>\n",
       "      <td>28230319</td>\n",
       "      <td>rs1800407</td>\n",
       "      <td>T</td>\n",
       "      <td>1.40</td>\n",
       "      <td>0.87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>92773663</td>\n",
       "      <td>92773664</td>\n",
       "      <td>rs12896399</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.58</td>\n",
       "      <td>-0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>33951693</td>\n",
       "      <td>33951694</td>\n",
       "      <td>rs16891982</td>\n",
       "      <td>C</td>\n",
       "      <td>-1.30</td>\n",
       "      <td>-0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11</td>\n",
       "      <td>89011046</td>\n",
       "      <td>89011047</td>\n",
       "      <td>rs1393350</td>\n",
       "      <td>A</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr      pos1      pos2          id minor_allele    b1    b2\n",
       "0   15  28365618  28365619  rs12913832            A -4.81 -1.79\n",
       "1   15  28230318  28230319   rs1800407            T  1.40  0.87\n",
       "2   14  92773663  92773664  rs12896399            G -0.58 -0.03\n",
       "3    5  33951693  33951694  rs16891982            C -1.30 -0.50\n",
       "4   11  89011046  89011047   rs1393350            A  0.47  0.27"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IRISPLEX = os.path.join(os.environ[\"HOME\"], \"project/datasets/\", \"irisplex.bed\")\n",
    "iris_header = [\"chr\", \"pos1\", \"pos2\", \"id\", \"minor_allele\", \"b1\", \"b2\"]\n",
    "iris = pd.read_csv(IRISPLEX, sep='\\t', header=None, names=iris_header)\n",
    "iris.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rs12913832',\n",
       " 'rs1800407',\n",
       " 'rs12896399',\n",
       " 'rs16891982',\n",
       " 'rs1393350',\n",
       " 'rs12203592']"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iris_snps = iris[\"id\"].values.tolist()\n",
    "iris_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gt = gt_nums\n",
    "gts = pd.read_csv(\"openSNP_final_genotypes.tab\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>6</th>\n",
       "      <th>8</th>\n",
       "      <th>2056</th>\n",
       "      <th>4106</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>13</th>\n",
       "      <th>...</th>\n",
       "      <th>2018</th>\n",
       "      <th>6117</th>\n",
       "      <th>2023</th>\n",
       "      <th>2024</th>\n",
       "      <th>6120</th>\n",
       "      <th>6124</th>\n",
       "      <th>4080</th>\n",
       "      <th>8178</th>\n",
       "      <th>6131</th>\n",
       "      <th>REF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82354</th>\n",
       "      <td>rs16891982</td>\n",
       "      <td>5</td>\n",
       "      <td>33987450</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95430</th>\n",
       "      <td>rs12203592</td>\n",
       "      <td>6</td>\n",
       "      <td>341321</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181970</th>\n",
       "      <td>rs1393350</td>\n",
       "      <td>11</td>\n",
       "      <td>88650694</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218381</th>\n",
       "      <td>rs12896399</td>\n",
       "      <td>14</td>\n",
       "      <td>91843416</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220888</th>\n",
       "      <td>rs1800407</td>\n",
       "      <td>15</td>\n",
       "      <td>25903913</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220908</th>\n",
       "      <td>rs12913832</td>\n",
       "      <td>15</td>\n",
       "      <td>26039213</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 597 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              rsid  chromosome  position  6    8  2056  4106   10   11   13  \\\n",
       "82354   rs16891982           5  33987450  2  2.0   2.0   2.0  1.0  1.0  2.0   \n",
       "95430   rs12203592           6    341321  1  0.0   0.0   0.0  0.0  0.0  0.0   \n",
       "181970   rs1393350          11  88650694  0  0.0   0.0   2.0  1.0  0.0  1.0   \n",
       "218381  rs12896399          14  91843416  0  0.0   1.0   1.0  0.0  0.0  1.0   \n",
       "220888   rs1800407          15  25903913  0  1.0   0.0   2.0  0.0  0.0  0.0   \n",
       "220908  rs12913832          15  26039213  1  0.0   2.0   2.0  0.0  0.0  0.0   \n",
       "\n",
       "       ...   2018  6117  2023  2024  6120  6124  4080  8178  6131  REF  \n",
       "82354  ...    2.0   NaN   2.0   2.0   2.0   2.0   2.0   2.0   NaN    C  \n",
       "95430  ...    1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0    C  \n",
       "181970 ...    2.0   0.0   0.0   1.0   1.0   1.0   1.0   0.0   1.0    G  \n",
       "218381 ...    1.0   NaN   1.0   2.0   0.0   1.0   2.0   0.0   NaN    G  \n",
       "220888 ...    0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   1.0    C  \n",
       "220908 ...    0.0   1.0   2.0   2.0   1.0   1.0   2.0   1.0   1.0    A  \n",
       "\n",
       "[6 rows x 597 columns]"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iris_genotypes = gts[gts[\"rsid\"].isin(iris_snps)]\n",
    "iris_genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>6</th>\n",
       "      <th>8</th>\n",
       "      <th>2056</th>\n",
       "      <th>4106</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>13</th>\n",
       "      <th>...</th>\n",
       "      <th>2023</th>\n",
       "      <th>2024</th>\n",
       "      <th>6120</th>\n",
       "      <th>6124</th>\n",
       "      <th>4080</th>\n",
       "      <th>8178</th>\n",
       "      <th>6131</th>\n",
       "      <th>REF_x</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs16891982</td>\n",
       "      <td>5</td>\n",
       "      <td>33987450</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>rs16891982</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs12203592</td>\n",
       "      <td>6</td>\n",
       "      <td>341321</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>C</td>\n",
       "      <td>rs12203592</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1393350</td>\n",
       "      <td>11</td>\n",
       "      <td>88650694</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>G</td>\n",
       "      <td>rs1393350</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs12896399</td>\n",
       "      <td>14</td>\n",
       "      <td>91843416</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>rs12896399</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1800407</td>\n",
       "      <td>15</td>\n",
       "      <td>25903913</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>C</td>\n",
       "      <td>rs1800407</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs12913832</td>\n",
       "      <td>15</td>\n",
       "      <td>26039213</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A</td>\n",
       "      <td>rs12913832</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 599 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         rsid  chromosome  position  6    8  2056  4106   10   11   13  ...    \\\n",
       "0  rs16891982           5  33987450  2  2.0   2.0   2.0  1.0  1.0  2.0  ...     \n",
       "1  rs12203592           6    341321  1  0.0   0.0   0.0  0.0  0.0  0.0  ...     \n",
       "2   rs1393350          11  88650694  0  0.0   0.0   2.0  1.0  0.0  1.0  ...     \n",
       "3  rs12896399          14  91843416  0  0.0   1.0   1.0  0.0  0.0  1.0  ...     \n",
       "4   rs1800407          15  25903913  0  1.0   0.0   2.0  0.0  0.0  0.0  ...     \n",
       "5  rs12913832          15  26039213  1  0.0   2.0   2.0  0.0  0.0  0.0  ...     \n",
       "\n",
       "   2023  2024  6120  6124  4080  8178  6131  REF_x          ID  REF_y  \n",
       "0   2.0   2.0   2.0   2.0   2.0   2.0   NaN      C  rs16891982      C  \n",
       "1   0.0   0.0   0.0   0.0   0.0   0.0   1.0      C  rs12203592      C  \n",
       "2   0.0   1.0   1.0   1.0   1.0   0.0   1.0      G   rs1393350      G  \n",
       "3   1.0   2.0   0.0   1.0   2.0   0.0   NaN      G  rs12896399      G  \n",
       "4   0.0   0.0   0.0   0.0   0.0   1.0   1.0      C   rs1800407      C  \n",
       "5   2.0   2.0   1.0   1.0   2.0   1.0   1.0      A  rs12913832      A  \n",
       "\n",
       "[6 rows x 599 columns]"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iris_merged = iris_genotypes.merge(oneK_ref, left_on=\"rsid\", right_on=\"ID\")\n",
    "iris_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_iris = iris_genotypes.merge(iris, left_on=\"rsid\", right_on=\"id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>6</th>\n",
       "      <th>8</th>\n",
       "      <th>2056</th>\n",
       "      <th>4106</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>13</th>\n",
       "      <th>...</th>\n",
       "      <th>8178</th>\n",
       "      <th>6131</th>\n",
       "      <th>REF</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos1</th>\n",
       "      <th>pos2</th>\n",
       "      <th>id</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>b1</th>\n",
       "      <th>b2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs16891982</td>\n",
       "      <td>5</td>\n",
       "      <td>33987450</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>5</td>\n",
       "      <td>33951693</td>\n",
       "      <td>33951694</td>\n",
       "      <td>rs16891982</td>\n",
       "      <td>C</td>\n",
       "      <td>-1.30</td>\n",
       "      <td>-0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs12203592</td>\n",
       "      <td>6</td>\n",
       "      <td>341321</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>396321</td>\n",
       "      <td>396322</td>\n",
       "      <td>rs12203592</td>\n",
       "      <td>T</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1393350</td>\n",
       "      <td>11</td>\n",
       "      <td>88650694</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>G</td>\n",
       "      <td>11</td>\n",
       "      <td>89011046</td>\n",
       "      <td>89011047</td>\n",
       "      <td>rs1393350</td>\n",
       "      <td>A</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs12896399</td>\n",
       "      <td>14</td>\n",
       "      <td>91843416</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>14</td>\n",
       "      <td>92773663</td>\n",
       "      <td>92773664</td>\n",
       "      <td>rs12896399</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.58</td>\n",
       "      <td>-0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1800407</td>\n",
       "      <td>15</td>\n",
       "      <td>25903913</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>C</td>\n",
       "      <td>15</td>\n",
       "      <td>28230318</td>\n",
       "      <td>28230319</td>\n",
       "      <td>rs1800407</td>\n",
       "      <td>T</td>\n",
       "      <td>1.40</td>\n",
       "      <td>0.87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs12913832</td>\n",
       "      <td>15</td>\n",
       "      <td>26039213</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A</td>\n",
       "      <td>15</td>\n",
       "      <td>28365618</td>\n",
       "      <td>28365619</td>\n",
       "      <td>rs12913832</td>\n",
       "      <td>A</td>\n",
       "      <td>-4.81</td>\n",
       "      <td>-1.79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 604 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         rsid  chromosome  position  6    8  2056  4106   10   11   13  ...   \\\n",
       "0  rs16891982           5  33987450  2  2.0   2.0   2.0  1.0  1.0  2.0  ...    \n",
       "1  rs12203592           6    341321  1  0.0   0.0   0.0  0.0  0.0  0.0  ...    \n",
       "2   rs1393350          11  88650694  0  0.0   0.0   2.0  1.0  0.0  1.0  ...    \n",
       "3  rs12896399          14  91843416  0  0.0   1.0   1.0  0.0  0.0  1.0  ...    \n",
       "4   rs1800407          15  25903913  0  1.0   0.0   2.0  0.0  0.0  0.0  ...    \n",
       "5  rs12913832          15  26039213  1  0.0   2.0   2.0  0.0  0.0  0.0  ...    \n",
       "\n",
       "   8178  6131  REF  chr      pos1      pos2          id  minor_allele    b1  \\\n",
       "0   2.0   NaN    C    5  33951693  33951694  rs16891982             C -1.30   \n",
       "1   0.0   1.0    C    6    396321    396322  rs12203592             T  0.70   \n",
       "2   0.0   1.0    G   11  89011046  89011047   rs1393350             A  0.47   \n",
       "3   0.0   NaN    G   14  92773663  92773664  rs12896399             G -0.58   \n",
       "4   1.0   1.0    C   15  28230318  28230319   rs1800407             T  1.40   \n",
       "5   1.0   1.0    A   15  28365618  28365619  rs12913832             A -4.81   \n",
       "\n",
       "     b2  \n",
       "0 -0.50  \n",
       "1  0.73  \n",
       "2  0.27  \n",
       "3 -0.03  \n",
       "4  0.87  \n",
       "5 -1.79  \n",
       "\n",
       "[6 rows x 604 columns]"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_iris"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_genotypes(x):\n",
    "    if x[\"rsid\"] in ['rs12896399', 'rs12913832', 'rs16891982']:\n",
    "        for col in x.index[3:-8]:\n",
    "            x[col] = 2 - x[col]\n",
    "    return x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>6</th>\n",
       "      <th>8</th>\n",
       "      <th>2056</th>\n",
       "      <th>4106</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>13</th>\n",
       "      <th>...</th>\n",
       "      <th>8178</th>\n",
       "      <th>6131</th>\n",
       "      <th>REF</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos1</th>\n",
       "      <th>pos2</th>\n",
       "      <th>id</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>b1</th>\n",
       "      <th>b2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs16891982</td>\n",
       "      <td>5</td>\n",
       "      <td>33987450</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>5</td>\n",
       "      <td>33951693</td>\n",
       "      <td>33951694</td>\n",
       "      <td>rs16891982</td>\n",
       "      <td>C</td>\n",
       "      <td>-1.30</td>\n",
       "      <td>-0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs12203592</td>\n",
       "      <td>6</td>\n",
       "      <td>341321</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>396321</td>\n",
       "      <td>396322</td>\n",
       "      <td>rs12203592</td>\n",
       "      <td>T</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1393350</td>\n",
       "      <td>11</td>\n",
       "      <td>88650694</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>G</td>\n",
       "      <td>11</td>\n",
       "      <td>89011046</td>\n",
       "      <td>89011047</td>\n",
       "      <td>rs1393350</td>\n",
       "      <td>A</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs12896399</td>\n",
       "      <td>14</td>\n",
       "      <td>91843416</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>14</td>\n",
       "      <td>92773663</td>\n",
       "      <td>92773664</td>\n",
       "      <td>rs12896399</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.58</td>\n",
       "      <td>-0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1800407</td>\n",
       "      <td>15</td>\n",
       "      <td>25903913</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>C</td>\n",
       "      <td>15</td>\n",
       "      <td>28230318</td>\n",
       "      <td>28230319</td>\n",
       "      <td>rs1800407</td>\n",
       "      <td>T</td>\n",
       "      <td>1.40</td>\n",
       "      <td>0.87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs12913832</td>\n",
       "      <td>15</td>\n",
       "      <td>26039213</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A</td>\n",
       "      <td>15</td>\n",
       "      <td>28365618</td>\n",
       "      <td>28365619</td>\n",
       "      <td>rs12913832</td>\n",
       "      <td>A</td>\n",
       "      <td>-4.81</td>\n",
       "      <td>-1.79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 604 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         rsid  chromosome  position  6    8  2056  4106   10   11   13  ...   \\\n",
       "0  rs16891982           5  33987450  0  0.0   0.0   0.0  1.0  1.0  0.0  ...    \n",
       "1  rs12203592           6    341321  1  0.0   0.0   0.0  0.0  0.0  0.0  ...    \n",
       "2   rs1393350          11  88650694  0  0.0   0.0   2.0  1.0  0.0  1.0  ...    \n",
       "3  rs12896399          14  91843416  2  2.0   1.0   1.0  2.0  2.0  1.0  ...    \n",
       "4   rs1800407          15  25903913  0  1.0   0.0   2.0  0.0  0.0  0.0  ...    \n",
       "5  rs12913832          15  26039213  1  2.0   0.0   0.0  2.0  2.0  2.0  ...    \n",
       "\n",
       "   8178  6131  REF  chr      pos1      pos2          id  minor_allele    b1  \\\n",
       "0   0.0   NaN    C    5  33951693  33951694  rs16891982             C -1.30   \n",
       "1   0.0   1.0    C    6    396321    396322  rs12203592             T  0.70   \n",
       "2   0.0   1.0    G   11  89011046  89011047   rs1393350             A  0.47   \n",
       "3   2.0   NaN    G   14  92773663  92773664  rs12896399             G -0.58   \n",
       "4   1.0   1.0    C   15  28230318  28230319   rs1800407             T  1.40   \n",
       "5   1.0   1.0    A   15  28365618  28365619  rs12913832             A -4.81   \n",
       "\n",
       "     b2  \n",
       "0 -0.50  \n",
       "1  0.73  \n",
       "2  0.27  \n",
       "3 -0.03  \n",
       "4  0.87  \n",
       "5 -1.79  \n",
       "\n",
       "[6 rows x 604 columns]"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_iris_fixed = final_iris.apply(fix_genotypes, 1)\n",
    "final_iris_fixed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code to compute probabilities for all individuals\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "# Functions for model probability calculations\n",
    "def p_blue(m1_sum, m2_sum):\n",
    "    return np.exp(m1_sum)/(1+np.exp(m1_sum)+np.exp(m2_sum))\n",
    "\n",
    "def p_other(m1_sum, m2_sum):\n",
    "    return np.exp(m2_sum)/(1+np.exp(m1_sum)+np.exp(m2_sum))\n",
    "\n",
    "# Perform predictions for each individual based on genotype and parameters\n",
    "a1 = 3.94\n",
    "a2 = 0.65\n",
    "\n",
    "predictions = {}\n",
    "for col in final_iris_fixed.columns[3:-10]:\n",
    "    pred1 = np.dot(final_iris_fixed[col], final_iris_fixed[\"b1\"]) + a1\n",
    "    pred2 = np.dot(final_iris_fixed[col], final_iris_fixed[\"b2\"]) + a2\n",
    "    blue = p_blue(pred1, pred2)\n",
    "    other = p_other(pred1, pred2)\n",
    "    brown = 1 - blue - other\n",
    "    predictions[col] = [pred1, pred2, blue, other, brown]\n",
    "\n",
    "predict = pd.DataFrame.from_dict(predictions, orient='index', columns=[\"pred1\", \"pred2\", \"blue\", \n",
    "                                                                       \"other\", \"brown\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_color_pred(x):\n",
    "    if x[\"blue\"] >= x[\"brown\"] and x[\"blue\"] >= x[\"other\"]:\n",
    "        return \"blue\"\n",
    "    elif x[\"brown\"] >= x[\"other\"]:\n",
    "        return \"brown\"\n",
    "    else:\n",
    "        return \"other\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "predict[\"predicted_eye_color\"] = predict.apply(get_color_pred, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "brown    242\n",
       "blue     231\n",
       "other    118\n",
       "Name: predicted_eye_color, dtype: int64"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predict[\"predicted_eye_color\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>pred1</th>\n",
       "      <th>pred2</th>\n",
       "      <th>blue</th>\n",
       "      <th>other</th>\n",
       "      <th>brown</th>\n",
       "      <th>predicted_eye_color</th>\n",
       "      <th>user_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6</td>\n",
       "      <td>-1.33</td>\n",
       "      <td>-0.47</td>\n",
       "      <td>0.139974</td>\n",
       "      <td>0.330780</td>\n",
       "      <td>0.529246</td>\n",
       "      <td>brown</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>-5.44</td>\n",
       "      <td>-2.12</td>\n",
       "      <td>0.003859</td>\n",
       "      <td>0.106754</td>\n",
       "      <td>0.889386</td>\n",
       "      <td>brown</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2056</td>\n",
       "      <td>3.36</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.909665</td>\n",
       "      <td>0.058737</td>\n",
       "      <td>0.031597</td>\n",
       "      <td>blue</td>\n",
       "      <td>2056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4106</td>\n",
       "      <td>7.10</td>\n",
       "      <td>2.90</td>\n",
       "      <td>0.984426</td>\n",
       "      <td>0.014762</td>\n",
       "      <td>0.000812</td>\n",
       "      <td>blue</td>\n",
       "      <td>4106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>-7.67</td>\n",
       "      <td>-3.22</td>\n",
       "      <td>0.000448</td>\n",
       "      <td>0.038403</td>\n",
       "      <td>0.961149</td>\n",
       "      <td>brown</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  index  pred1  pred2      blue     other     brown predicted_eye_color  \\\n",
       "0     6  -1.33  -0.47  0.139974  0.330780  0.529246               brown   \n",
       "1     8  -5.44  -2.12  0.003859  0.106754  0.889386               brown   \n",
       "2  2056   3.36   0.62  0.909665  0.058737  0.031597                blue   \n",
       "3  4106   7.10   2.90  0.984426  0.014762  0.000812                blue   \n",
       "4    10  -7.67  -3.22  0.000448  0.038403  0.961149               brown   \n",
       "\n",
       "   user_id  \n",
       "0        6  \n",
       "1        8  \n",
       "2     2056  \n",
       "3     4106  \n",
       "4       10  "
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictions = predict.reset_index()\n",
    "predictions['user_id'] = predictions['index'].astype(int)\n",
    "predictions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "brown    242\n",
       "blue     230\n",
       "other      8\n",
       "Name: predicted_eye_color, dtype: int64"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_nona[\"predicted_eye_color\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = phenotypes.merge(predictions, on=\"user_id\")\n",
    "result_nona = result.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    }
   ],
   "source": [
    "result_nona[\"pred_correct\"] = result_nona[\"eye_color\"].str.lower().values == result_nona[\"predicted_eye_color\"].str.lower().values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>user_id</th>\n",
       "      <th>pred1</th>\n",
       "      <th>pred2</th>\n",
       "      <th>blue</th>\n",
       "      <th>other</th>\n",
       "      <th>brown</th>\n",
       "      <th>pred_correct</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eye_color</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>blue</th>\n",
       "      <td>3898.638037</td>\n",
       "      <td>4.075276</td>\n",
       "      <td>1.188957</td>\n",
       "      <td>0.914405</td>\n",
       "      <td>0.058675</td>\n",
       "      <td>0.026920</td>\n",
       "      <td>0.98773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>brown</th>\n",
       "      <td>3764.271255</td>\n",
       "      <td>-3.344615</td>\n",
       "      <td>-1.643117</td>\n",
       "      <td>0.119612</td>\n",
       "      <td>0.167862</td>\n",
       "      <td>0.712526</td>\n",
       "      <td>0.94332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>green</th>\n",
       "      <td>4333.642857</td>\n",
       "      <td>2.657286</td>\n",
       "      <td>0.639571</td>\n",
       "      <td>0.764856</td>\n",
       "      <td>0.124405</td>\n",
       "      <td>0.110740</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               user_id     pred1     pred2      blue     other     brown  \\\n",
       "eye_color                                                                  \n",
       "blue       3898.638037  4.075276  1.188957  0.914405  0.058675  0.026920   \n",
       "brown      3764.271255 -3.344615 -1.643117  0.119612  0.167862  0.712526   \n",
       "green      4333.642857  2.657286  0.639571  0.764856  0.124405  0.110740   \n",
       "\n",
       "           pred_correct  \n",
       "eye_color                \n",
       "blue            0.98773  \n",
       "brown           0.94332  \n",
       "green           0.00000  "
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_nona.groupby(\"eye_color\").mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8208333333333333"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean(result_nona[\"eye_color\"].str.lower().values == result_nona[\"predicted_eye_color\"].str.lower().values)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
